File: samples/dark_castle.ts
/*
 * In a land shrouded in mystery ...
 */

interface DarkCastleProps {
  name: string;
  constructionYear: number;
  destructionYear: number;
  features: string[];
}

class DarkCastle {
  name: string;
  constructionYear: number;
  destructionYear: number;
  features: string[];

  constructor(props: DarkCastleProps) {
    this.name = props.name;
    this.constructionYear = props.constructionYear;
    this.destructionYear = props.destructionYear;
    this.features = props.features;
  }

  get duration(): number {
    return this.calcDuration();
  }

  calcDuration(): number {
    return this.destructionYear - this.constructionYear;
  }
}

// ...there stood a Dark Castle

const MoonshadowFortress = new DarkCastle({
  name: "Moonshadow Fortress",
  constructionYear: 980,
  destructionYear: 1503,
  features: ["Eerie Glow", "High Battlements"],
});

console.log(`The duration of ${MoonshadowFortress.name} was ${MoonshadowFortress.duration} years.`);

File: samples/dark_castle.kt
/*
 * In a land shrouded in mystery...
 */

class DarkCastle(
  val name: String,
  val constructionYear: Int,
  val destructionYear: Int,
  val features: Array<String>
) {
  val duration: Int
    get() = this.calcDuration()

  fun calcDuration() =
    this.destructionYear - this.constructionYear
}

// ...there stood a Dark Castle

fun main() {
  val castle = DarkCastle(
    "Moonshadow Fortress",
    980,
    1503,
    arrayOf("Eerie Glow", "High Battlements")
  )

  println("The duration of ${castle.name} was ${castle.duration} years.")
}
